examples/ota-requestor-app/esp32/main/OTARequesterImpl.cpp

```diff
@@ -0,0 +1,282 @@
+using chip::NodeId;
+using chip::Server;
+using chip::VendorId;
+using chip::bdx::TransferSession;
+using chip::Callback::Callback;
+using chip::Inet::IPAddress;
+using chip::System::Layer;
+using chip::Transport::PeerAddress;
+using namespace chip::Messaging;
+using namespace chip::app::device;
+
+void OnQueryImageResponse(void * context, uint8_t status, uint32_t delayedActionTime, CharSpan imageURI, uint32_t softwareVersion,
+                          CharSpan softwareVersionString, ByteSpan updateToken, bool userConsentNeeded,
+                          ByteSpan metadataForRequester);
+void OnQueryImageFailure(void * context, uint8_t status);
+
+void OnApplyUpdateResponse(void * context, uint8_t action, uint32_t delayedActionTime);
+void OnApplyUpdateRequestFailure(void * context, uint8_t status);
+
+void OnConnected(void * context, OperationalDeviceProxy * operationalDeviceProxy);
+void OnConnectionFailure(void * context, OperationalDeviceProxy * operationalDeviceProxy, CHIP_ERROR error);
+
+void OnBlockReceived(void * context, chip::bdx::TransferSession::BlockData & blockdata);
+void OnTransferComplete(void * context);
+void OnTransferFailed(void * context);
+
+// TODO: Encapsulate these globals and the callbacks in some class
+OperationalDeviceProxy gOperationalDeviceProxy;
+ExchangeContext * exchangeCtx = nullptr;
+BdxDownloader bdxDownloader;
+uint8_t operationalDeviceContext; // 0 - none, 1 - QueryImage, 2 - ApplyUpdateRequest, 3 - NotifyUpdateApplied
+ByteSpan otaUpdateToken;
+
+/* Callbacks for QueryImage response */
+Callback<OtaSoftwareUpdateProviderClusterQueryImageResponseCallback> mQueryImageResponseCallback(OnQueryImageResponse, nullptr);
+Callback<DefaultFailureCallback> mOnQueryImageFailureCallback(OnQueryImageFailure, nullptr);
+
+/* Callbacks for ApplyUpdateRequest response */
+Callback<OtaSoftwareUpdateProviderClusterApplyUpdateRequestResponseCallback> mApplyUpdateResponseCallback(OnApplyUpdateResponse,
+                                                                                                          nullptr);
+Callback<DefaultFailureCallback> mOnApplyUpdateRequestFailureCallback(OnApplyUpdateRequestFailure, nullptr);
+
+/* Callbacks for operational device proxy connect response */
+Callback<OnOperationalDeviceConnected> mOnConnectedCallback(OnConnected, &operationalDeviceContext);
+Callback<OnOperationalDeviceConnectionFailure> mOnConnectionFailureCallback(OnConnectionFailure, nullptr);
+
+/* Callbacks for BDX data transfer */
+Callback<OnBdxBlockReceived> mOnBlockReceived(OnBlockReceived, nullptr);
+Callback<OnBdxTransferComplete> mOnTransferComplete(OnTransferComplete, nullptr);
+Callback<OnBdxTransferFailed> mOnTransferFailed(OnTransferFailed, nullptr);
+
+FabricIndex providerFabricIndex = 1;
+uint16_t setupDiscriminator     = CHIP_DEVICE_CONFIG_USE_TEST_SETUP_DISCRIMINATOR;
+
+void OnQueryImageResponse(void * context, uint8_t status, uint32_t delayedActionTime, CharSpan imageURI, uint32_t softwareVersion,
```

Comment:
be aware you may need to do some refactor after https://github.com/project-chip/connectedhomeip/pull/11336

Reply:
Rebased!